core/main/src/firebolt/handlers/hdmi_rpc.rs

```diff
@@ -0,0 +1,198 @@
+    ) -> RpcResult<ListenerResponse>;
+}
+
+#[derive(Debug)]
+pub struct HdmiImpl {
+    pub state: PlatformState,
+}
+
+#[async_trait]
+impl HdmiServer for HdmiImpl {
+    async fn select_hdmi_operation(
+        &self,
+        _ctx: CallContext,
+        request: HdmiSelectOperationRequest,
+    ) -> RpcResult<HdmiSelectOperationResponse> {
+        if let HdmiOperation::Start = request.operation {
+            if let Ok(response) = self
+                .state
+                .get_client()
+                .send_extn_request(HdmiRequest::StartHdmiInput(request))
+                .await
+            {
+                if let ExtnPayload::Response(ExtnResponse::Value(value)) = response.payload {
+                    if let Ok(res) = serde_json::from_value::<HdmiSelectOperationResponse>(value) {
+                        return Ok(res);
+                    }
+                }
+            }
+
+            Err(rpc_err("FB error response TBD"))
+        } else {
+            if let Ok(response) = self
+                .state
+                .get_client()
+                .send_extn_request(HdmiRequest::StopHdmiInput(request))
+                .await
+            {
+                if let ExtnPayload::Response(ExtnResponse::Value(value)) = response.payload {
+                    if let Ok(res) = serde_json::from_value::<HdmiSelectOperationResponse>(value) {
+                        return Ok(res);
+                    }
+                }
+            }
+
+            Err(rpc_err("FB error response TBD"))
+        }
+    }
+
+    async fn get_available_inputs(
+        &self,
+        _ctx: CallContext,
+    ) -> RpcResult<GetAvailableInputsResponse> {
+        if let Ok(response) = self
+            .state
+            .get_client()
+            .send_extn_request(HdmiRequest::GetAvailableInputs)
+            .await
+        {
+            if let ExtnPayload::Response(ExtnResponse::Value(value)) = response.payload {
+                if let Ok(res) = serde_json::from_value::<GetAvailableInputsResponse>(value) {
+                    return Ok(res);
+                }
+            }
+        }
+
+        Err(rpc_err("FB error response TBD"))
+    }
+    async fn on_hdmi_connection_changed(
+        &self,
+        ctx: CallContext,
+        request: ListenRequest,
+    ) -> RpcResult<ListenerResponse> {
+        let listen = request.listen;
+
+        AppEvents::add_listener(
+            &self.state,
+            HDMI_CONNECTION_CHANGED.to_string(),
+            ctx.clone(),
+            request,
+        );
+
+        if self
+            .state
+            .get_client()
+            .send_extn_request(DeviceEventRequest {
+                event: DeviceEvent::HdmiConnectionChanged,
+                id: ctx.app_id,
+                subscribe: listen,
+                callback_type: DeviceEventCallback::FireboltAppEvent,
+            })
+            .await
+            .is_err()
+        {
+            error!("Error while registration");
```

Comment:
Should we still return an OK response even if we hit this block?

Reply:
This is similar to how on_hdcp_changed listener is implemented https://github.com/rdkcentral/Ripple/blob/0565eeeb871988ca1d4a8535d83ce9dc50e3d6e6/core/main/src/firebolt/handlers/device_rpc.rs#L417

Can you elaborate on how I should change it?

Reply:
Got it. thank you :)

Reply:
Looks like there is a util function `rpc_error` that can be used to glean it up a bit:
```
return Err(Error::Custom("Failed to subscribe to HdmiConnectionChanged event".to_owned()));
```

Becomes:
```
return Err(rpc_error("Failed to subscribe to HdmiConnectionChanged event"));
```

Sorry, I didn't know about this fn until just now.